cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -907,21 +986,189 @@ class CastNode extends Node {
+
+  override Location getLocation() { result = sc.getLocation() }
+}
 
 class DataFlowExpr = Expr;
 
 class DataFlowType = Type;
 
-/** A function call relevant for data flow. */
-class DataFlowCall extends CallInstruction {
-  DataFlowCallable getEnclosingCallable() { result = this.getEnclosingFunction() }
+cached
+newtype TDataFlowCall =
+  TNormalCall(CallInstruction call) or
+  TSummaryCall(
+    FlowSummaryImpl::Public::SummarizedCallable c, FlowSummaryImpl::Private::SummaryNode receiver
+  ) {
+    FlowSummaryImpl::Private::summaryCallbackRange(c, receiver)
+  }
+
+/**
+ * A function call relevant for data flow. This includes calls from source
+ * code and calls inside library callables with a flow summary.
+ */
+class DataFlowCall extends TDataFlowCall {
+  /**
+   * Gets the underlying data flow call instruction, if any.
+   */
+  CallInstruction asCallInstruction() { none() }
+
+  /**
+   * Gets the operand the specifies the target function of the call.
+   */
+  CallTargetOperand getCallTargetOperand() { none() }
+
+  /**
+   * Gets the `Function` that the call targets, if this is statically known.
+   */
+  DataFlowCallable getStaticCallTarget() { none() }
+
+  /**
+   * Gets the `index`'th argument operand. The qualifier is considered to have index `-1`.
+   */
+  ArgumentOperand getArgumentOperand(int index) { none() }
+
+  /**
+   * Gets the argument at the specified index, or `this` if `index` is `-1`.
+   */
+  pragma[noinline]
+  final Instruction getArgument(int index) { result = this.getArgumentOperand(index).getDef() }
+
+  /**
+   * Gets the number of arguments of the call, including the `this` pointer, if any.
+   */
+  final int getNumberOfArguments() { result = count(this.getArgumentOperand(_)) }
+
+  /**
+   * Gets the enclosing callable, if any.
+   */
+  DataFlowCallable getEnclosingCallable() { none() }
+
+  /**
+   * Gets a textual representation of this call.
+   */
+  string toString() { none() }
+
+  /**
+   * Gets the location of this call.
+   */
+  Location getLocation() { none() }
+}
+
+/**
+ * TODO: QLDoc.
+ */
+private class NormalCall extends DataFlowCall, TNormalCall {
+  private CallInstruction call;
+
+  NormalCall() { this = TNormalCall(call) }
+
+  override CallInstruction asCallInstruction() { result = call }
+
+  override CallTargetOperand getCallTargetOperand() { result = call.getCallTargetOperand() }
+
+  override DataFlowCallable getStaticCallTarget() {
+    result = TSourceCallable(call.getStaticCallTarget())
+  }
+
+  override ArgumentOperand getArgumentOperand(int index) { result = call.getArgumentOperand(index) }
+
+  override DataFlowCallable getEnclosingCallable() {
+    result = TSourceCallable(call.getEnclosingFunction())
+  }
+
+  override string toString() { result = call.toString() }
+
+  override Location getLocation() { result = call.getLocation() }
+}
+
+/**
+ * A synthesized call inside a callable with a flow summary.
+ *
+ * TODO: example.
```

Comment:
I'm not sure when a `SummaryCall` is generated.  I don't appear to have any in my tests.

Reply:
A `SummaryCall` is a call that happens inside a modelled function. For example, if we had a model for:
```cpp
void call_argument(void (*f)()) {
  f();
}
```
there will be a `SummaryCall` that represents the call to `f`.

Reply:
Can we model that call - or does this only apply when the function is both modelled and implemented?

Reply:
We can model that call, yeah. Let's give it a parameter so that there's some interesting flow:

```cpp
void call_argument_2(int x, void (*f)(int)) {
  f(x);
}
```
We can model the behavior of `call_argument_2` with a MaD row like:
```
;;false;call_argument_2;;;Argument[0];Argument[1].Parameter[0];value
```
that is, there is flow from `Argument[0]` to the 0'th parameter of the 1st function. So if there's a declaration such as:
```
void call_argument_2(int x, void (*f)(int));
```
without any definition, then a MaD model like the above should ensure that we have flow from `x` to the callback's parameter.

Reply:
Ah, this is quite similar to some of my existing test cases - and it appears they *do* in fact generate `SummaryCall`s, but I was not testing for them properly.

I've just pushed a new test + QLDoc for `SummaryCall`, `SourceCallable` and `SummarizedCallable`.  Please check the QLDoc in particular is accurate, I'm happy to accept corrections and even minor clarifications as I depend on these comments for my own understanding.